docs: add documentation for using lambda functions (#7092)
confluentinc/ksql: docs/developer-guide/ksqldb-reference/scalar-functions.md

Slices a list based on the supplied indices. The indices start at 1 and
include both endpoints.

## Invocation Functions

> Used to apply lambda functions to collections.

Reviewer: ```suggestion
Apply lambda functions to collections.
```

### `TRANSFORM`

Since: 0.17.0

```sql
TRANSFORM(array, x => ...)

TRANSFORM(map, (k,v) => ..., (k,v) => ...)
```

> Transform a collection with a lambda function.

Reviewer: ```suggestion
Transform a collection by using a lambda function.
```

> If the collection is an array,  the lambda function must have 1 input argument.

Reviewer: ```suggestion
If the collection is an array, the lambda function must have one input argument.
```

> If the collection is a map, two lambda functions must be provided and each lambda function must have 2 arguments (map entry key and map entry value).

Reviewer: ```suggestion
If the collection is a map, two lambda functions must be provided, and both lambdas must have two arguments: a map entry key and a map entry value.
```

### `Reduce`

Since: 0.17.0

```sql
REDUCE(state, array, (s, x) => ...)

> REDUCE(state, map, (x, k, v) => ...)

Author: This should be an `s` instead of an `x` -> I'll update

```

Reduce a collection starting from an initial state.

> If the collection is an array, the lambda function must have 2 input arguments.

Reviewer: ```suggestion
If the collection is an array, the lambda function must have two input arguments.
```

> If the collection is a map, the lambda function must have 3 input arguments.

Reviewer: ```suggestion
If the collection is a map, the lambda function must have three input arguments.
```

> If the state is `null` the result will be `null`.

Reviewer: ```suggestion
If the state is `null`, the result is `null`.
```

### `Filter`

Since: 0.17.0

```sql
FILTER(array, x => ...)

FILTER(map, (k,v) => ...)
```

Filter a collection with a lambda function.

> If the collection is an array, the lambda function must have 1 input argument.

Reviewer: ```suggestion
If the collection is an array, the lambda function must have one input argument.
```

> If the collection is a map, the lambda function must have 2 input arguments.

Reviewer: ```suggestion
If the collection is a map, the lambda function must have two input arguments.
```
